### PR TITLE
Setup basic webapp for decoding CoAP messages

### PR DESCRIPTION
**Pull request by twyatt**
Partial (decoding only) implementation for #41.

GitHub Actions updates:
- Split CI and publish/snapshot actions
- Deploy online decoder on merge to `main`

https://user-images.githubusercontent.com/98017/105253420-4b055600-5b34-11eb-94e2-bd7d2b9eaa2c.mov

---

To experiment with the webapp on your local machine, checkout branch then:

```
$ ./gradlew assembleGitHubPages
$ open build/gh-pages/index.html
```

### CODE REVIEW

**Comment by twyatt** [comment]
Note to self: update build scripts to deploy on merges to `main`.

File: webapp/src/main/resources/index.html [hunk 1]
```diff
@@ -0,0 +1,45 @@
+<!doctype html>
+<html lang="en" xmlns="http://www.w3.org/1999/html">
+    <head>
+        <meta charset="UTF-8">
```
Comment (cedrickcooke) [inline webapp/src/main/resources/index.html:4]: Fun fact this is one of my biggest pet peeves (not your fault, I realize it's the spec). How the hell is a text parser supposed to find `<meta charset="UTF-8">` if it wasn't already using the correct encoding? And why doesn't HTML close all tags `/>` 😭

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,27 +1,14 @@
-# Each virtual machine has the same hardware resources available.
-#
-# - 2-core CPU
-# - 7 GB of RAM memory
-# - 14 GB of SSD disk space
-#
-# https://help.github.com/en/actions/automating-your-workflow-with-github-actions/virtual-environments-for-github-hosted-runners
-
 name: CI
```
Comment (cedrickcooke) [inline .github/workflows/ci.yml:1]: Nit: we could go with a more descriptive name than just CI. "Branch Check"?

**Review by cedrickcooke — APPROVED** [review]

File: webapp/src/main/resources/index.html (hunk 1 above)
Comment (twyatt) [reply webapp/src/main/resources/index.html:4]: Haha. The web is a crazy-wild-beast.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (twyatt) [reply .github/workflows/ci.yml:1]: It runs on pushes (e.g. pushing a commit to `main` or even pushes to PR branches), so, technically it is checking a branch (but more the commits on that branch).

So, "branch check" is more specific, but I think CI is so engrained into me now it feels more right here?

File: webapp/src/main/kotlin/serialization/MessageSerializer.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    override fun deserialize(decoder: Decoder): Code = TODO("Not yet implemented")
+}
+
+private object OptionSerializer : KSerializer<Option> {
+
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Option", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: Option) {
+        val serialized = when (value) {
```
Comment (sdonn3) [inline webapp/src/main/kotlin/serialization/MessageSerializer.kt:130]: seems like the when here and below could be simplified into one?

Comment (twyatt) [reply webapp/src/main/kotlin/serialization/MessageSerializer.kt:130]: Tried a few things but couldn't figure out how to simplify it further?

Comment (sdonn3) [reply webapp/src/main/kotlin/serialization/MessageSerializer.kt:130]: might just be personal preference and thus I would never enforce it on ya but I think

`when (value) {
	Option.ContentFormat.PlainText -> "Content-Format:text/plain; charset=utf-8"
	...`

would be cleaner and easier to read?

**Review by sdonn3 — APPROVED** [review]

File: webapp/src/main/kotlin/serialization/MessageSerializer.kt (hunk 3 above)
Comment (twyatt) [reply webapp/src/main/kotlin/serialization/MessageSerializer.kt:137]: Oh, gotchya. Ya, I agree with that.

Unfortunately I snuck some other changes in to properly handle detecting encoding by also reading the `Accept` option. It muddies up the entire thing. So it has gotten slightly messy, but just being a debug tool I think that's acceptable.

**tl;dr** code has changed enough that original comment isn't really applicable anymore.
